node/src/components/chain_synchronizer/operations.rs

```diff
@@ -341,125 +347,196 @@ impl CanUseSyncingNodes for Deploy {}
+/// Gets a list of peers suitable for the fetch operation.
+async fn get_peers<REv>(include_syncing: bool, ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
 where
     REv: From<NetworkInfoRequest>,
 {
-    let mut peer_list = ctx
-        .effect_builder
-        .get_fully_connected_non_syncing_peers()
-        .await;
+    let mut peer_list = if include_syncing {
+        ctx.effect_builder.get_fully_connected_peers().await
+    } else {
+        ctx.effect_builder
+            .get_fully_connected_non_syncing_peers()
+            .await
+    };
     ctx.filter_bad_peers(&mut peer_list);
     peer_list
 }
 
-/// Returns fully-connected, syncing and non-syncing peers that are known to be not banned.
-async fn get_filtered_fully_connected_peers<REv>(ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
+/// Fetching should stop only when the network is connected and all retries have been exhausted.
+fn should_stop_fetching(
+    retry_count: &mut usize,
+    has_connected_to_network: bool,
+    max_sync_fetch_attempts: usize,
+) -> bool {
+    if !has_connected_to_network {
+        return false;
+    }
+
+    *retry_count += 1;
+    if *retry_count <= max_sync_fetch_attempts {
+        return false;
+    }
+
+    true
+}
+
+#[derive(Error, Debug)]
+pub(crate) enum FetchWithRetryError<T>
 where
-    REv: From<NetworkInfoRequest>,
+    T: Item,
 {
-    let mut peer_list = ctx.effect_builder.get_fully_connected_peers().await;
-    ctx.filter_bad_peers(&mut peer_list);
-    peer_list
+    #[error("Fetch retries exhausted for item with id {id:?}. Total attempts: {total_attempts}, retries while network was connected: {retries_while_connected}")]
+    RetriesExhausted {
+        id: T::Id,
+        total_attempts: usize,
+        retries_while_connected: usize,
+    },
+
+    #[error(transparent)]
+    FetcherError(#[from] FetcherError<T>),
 }
 
-/// Fetches an item. Keeps retrying to fetch until it is successful. Not suited to fetching a block
-/// header or block by height, which require verification with finality signatures.
-async fn fetch_retry_forever<REv, T>(ctx: &ChainSyncContext<'_, REv>, id: T::Id) -> FetchResult<T>
+/// Fetches an item. Not suited to fetching a block header or block by height, which require
+/// verification with finality signatures.
+async fn fetch_with_retries<REv, T>(
+    ctx: &ChainSyncContext<'_, REv>,
+    id: T::Id,
+) -> Result<FetchedData<T>, FetchWithRetryError<T>>
 where
     T: Item + CanUseSyncingNodes + 'static,
     REv: From<FetcherRequest<T>> + From<NetworkInfoRequest>,
 {
-    let mut attempts = 0_usize;
+    let mut total_attempts = 0_usize;
+    let mut retry_count = 0_usize;
     loop {
-        let new_peer_list = if T::can_use_syncing_nodes() {
-            get_filtered_fully_connected_peers(ctx).await
-        } else {
-            get_filtered_fully_connected_non_syncing_peers(ctx).await
-        };
+        let has_connected_to_network = has_connected_to_network();
+        if should_stop_fetching(
```

Comment:
If you look at the suggestion on how to shorten the function above, I am inclined to inline it entirely here. I think passing a mutable `&mut usize` is a bit of an issue, especially since the whole counting logic is now split across this function and `fetch_with_retries`.

I *do* appreciate the effort to make it testable though, so there's that. Another option would be to build a proper state machine and test that? Knowing how things went in the past, we will probably make the retry logic even more complex down the line :)

```rust
#[derive(Debug, Default)]
struct RetryOperation {
    is_connected: bool,
    attempts: usize,
    retries: usize,
    // not sure about the conditions exactly.
    max_attempts: usize,
    max_retries: usize,
}

impl RetryOperation {
    fn set_connected(&mut self, is_connected: bool) {
        self.is_connected = is_connected;
    }

    /// Prepare another attempt of the operation.
    ///
    /// If `false` is returned, do not actually perform an attempt, but sleep.
    fn attempt() -> bool {
        self.attempts += 1;
        // TODO: max attempts?
        self.is_connected
    }

    /// Notify that the operation was tried and failed.
    ///
    /// If `false` is returned, give up on the operation.
    fn retry() -> bool {
        // ... etc
    }
}
```

This could also use one generic struct or error that contains the info from the retry operation.

If we wanted to go further a closure `FnMut() -> impl Future<_>` could be passed in that performs one iteration of the attempts and all the logic shared, but that may be overkill.

Reply:
I also see the point in splitting this in such a way that it can be unit tested, but I don't feel that it's worth the additional complexity given how little of the retry logic actually gets unit test coverage.  That applies to the code as in the PR and also the suggestion to create a state machine struct above.

I'd think this function could be implemented reasonably cleanly as follows:
```rust
    let mut total_attempts = 0;
    let mut attempts_after_bootstrapped = 0;
    loop {
        let has_connected_to_network = has_connected_to_network();
        let new_peer_list = get_peers(T::can_use_syncing_nodes(), ctx).await;
        if new_peer_list.is_empty() && total_attempts % 100 == 0 {
            warn!(
                total_attempts,
                attempts_after_bootstrapped,
                item_type = ?T::TAG,
                ?id,
                can_use_syncing_nodes = %T::can_use_syncing_nodes(),
                has_connected_to_network,
                "failed to attempt to fetch item due to no fully-connected peers"
            );
        }

        if let Some(value) = fetch_from_peers(new_peer_list, id, ctx).await {
            return value;
        }

        total_attempts += 1;
        if has_connected_to_network {
            attempts_after_bootstrapped += 1;
        }
        if attempts_after_bootstrapped >= ctx.config.max_sync_fetch_attempts() {
            return Err(FetchWithRetryError::AttemptsExhausted {
                id,
                total_attempts,
                attempts_after_bootstrapped,
            });
        }

        tokio::time::sleep(ctx.config.retry_interval()).await;
    }
```

Reply:
I'd be happy with that as well. Whether or not to factor it out for better testability, I leave up to you :)

Reply:
Taken the approach proposed by @Fraser999 in bc168bc70f2f3c3d85a001a36556297be734e1db